### PR TITLE
feat(ingestion): Mode retry wait logic to avoid hitting Mode API rate limit

### PR DESCRIPTION
**Pull request by jawadqu**
Implement exponential backoff retry logic when accessing Mode API to avoid HTTP error 429: "Too many API request".
See Datahub Mode connector [Config Details](https://github.com/jawadqu/datahub/blob/62a0f5e9740fb7cc1cc157ae7fdcb06dbb479156/metadata-ingestion/source_docs/mode.md#config-details) on how to customize these settings.

## Checklist
- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [X] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Review by shirshanka — CHANGES_REQUESTED** [review]
small comment... LGTM otherwise.

File: metadata-ingestion/src/datahub/ingestion/source/mode.py [hunk 1]
```diff
@@ -169,12 +182,21 @@ def _get_ownership(self, user: str) -> Optional[OwnershipClass]:
...
-        user_json = user.json()
-        if self.config.owner_username_instead_of_email:
-            return user_json.get("username", "unknown")
-        else:
-            return user_json.get("email", "unknown")
+        user = ""
+        try:
+            user_json = self._get_request_json(f"{self.config.connect_uri}{href}")
+            user = (
+                user_json.get("username", "unknown")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/mode.py:189]: Nit: if we don't get a username, i'd rather we don't say anything as opposed to putting some "unknown" stuff in the DB

File: metadata-ingestion/src/datahub/ingestion/source/mode.py [hunk 2]
```diff
@@ -591,6 +615,36 @@ def _get_charts(self, report_token: str, query_token: str) -> list:
...
+        def get_request():
+            try:
+                response = self.session.get(url)
+                response.raise_for_status()
+                return response.json()
+            except HTTPError as http_error:
+                error_response = http_error.response
+                if error_response.status_code == 429:
+                    # respect Retry-After
+                    sleep_time = error_response.headers.get("retry-after")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/mode.py:638]: this is great!

File: metadata-ingestion/src/datahub/ingestion/source/mode.py [hunk 3]
```diff
@@ -591,6 +615,36 @@ def _get_charts(self, report_token: str, query_token: str) -> list:
...
+                    if sleep_time is not None:
+                        time.sleep(sleep_time)
+                    raise HTTPError429
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/mode.py:641]: awesome.

File: metadata-ingestion/src/datahub/ingestion/source/mode.py [hunk 4]
```diff
@@ -334,9 +354,19 @@ def _get_datahub_friendly_platform(self, adapter, platform):
     def _get_platform_and_dbname(
         self, data_source_id: int
     ) -> Union[Tuple[str, str], Tuple[None, None]]:
-        ds_response = self.session.get(f"{self.workspace_uri}/data_sources")
-        ds_json = ds_response.json()
-        data_sources = ds_json.get("_embedded", {}).get("data_sources", {})
+
+        data_sources = {}
```
[the remark this answers is not in the document]
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/mode.py]: fixed by [commit de60b3b826e1e2bf1a307ba8372ad4fa6f24d11e](https://github.com/linkedin/datahub/pull/3761/commits/de60b3b826e1e2bf1a307ba8372ad4fa6f24d11e)

File: metadata-ingestion/src/datahub/ingestion/source/mode.py (hunk 1 above)
Comment (jawadqu) [reply metadata-ingestion/src/datahub/ingestion/source/mode.py]: fixed by [commit 62afea1237a57c54b65845b1fe0a313a037f7c88](https://github.com/linkedin/datahub/pull/3761/commits/62afea1237a57c54b65845b1fe0a313a037f7c88).

creating an ChangeAuditStamps does create an URN with unknown name though.
`ChangeAuditStampsClass({'created': AuditStampClass({'time': 0, 'actor': 'urn:li:corpuser:unknown', 'impersonator': None}), 'lastModified': AuditStampClass({'time': 0, 'actor': 'urn:li:corpuser:unknown', 'impersonator': None}), 'deleted': None})`

**Review by shirshanka — APPROVED** [review]
LGTM

**Comment by shubhamg931** [comment]
Hey, do you know if we need to do any further optimizations here? We are getting this error repeatedly for updating around 10k datasets every 15 minutes for a 3 node `t3.small.search` OpenSearch cluster

Our version already have this fix, still not helping

**Pull request merged**